[SwiftUI] Add more of the main functionality to WebPage and WebView

rr-codes opened the pull request:
#### 408608a269ff1452e2d40ca05aeab6470c4d289b
<pre>
[SwiftUI] Add more of the main functionality to WebPage and WebView
<a href="https://bugs.webkit.org/show_bug.cgi?id=283653">https://bugs.webkit.org/show_bug.cgi?id=283653</a>
<a href="https://rdar.apple.com/140507710">rdar://140507710</a>

Reviewed by Abrar Rahman Protyasha.

This adds most of the missing capabilities to WebPage and WebView.

* Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift: Added.
(View.webViewAllowsBackForwardNavigationGestures(_:)):
(View.webViewAllowsLinkPreview(_:)):
(View.webViewAllowsTabFocusingLinks(_:)):
(View.webViewAllowsTextInteraction(_:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+FrameInfo.swift: Added.
(FrameInfo.isMainFrame):
(FrameInfo.request):
(FrameInfo.securityOrigin):
(FrameInfo.wrapped):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.isWritingToolsActive):
(WebPage_v0.handlesURLScheme(_:)):
(WebPage_v0.load(_:mimeType:characterEncoding:baseURL:)):
(WebPage_v0.load(_:allowingReadAccessTo:)):
(WebPage_v0.loadSimulatedRequest(_:response:responseData:)):
(WebPage_v0.loadSimulatedRequest(_:responseHTML:)):
(WebPage_v0.reload(_:)):
(WebPage_v0.stopLoading):
(WebPage_v0.callAsyncJavaScript(_:arguments:in:contentWorld:)):
(WebPage_v0.snapshot(_:)):
(WebPage_v0.pdf(_:)):
(WebPage_v0.webArchiveData):
* Source/WebKit/UIProcess/API/Swift/WebView.swift:
(WebViewRepresentable.updatePlatformView(_:context:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/MiniBrowser/SwiftUI/ContentView.swift:
(UnifiedBar.body):
(ContentView.body):
(ContentView.urlFieldIsFocused): Deleted.
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift:
(WebPageTests.javaScriptEvaluation):

Canonical link: <a href="https://commits.webkit.org/287428@main">https://commits.webkit.org/287428@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d8a80c9e74fdf0b534b5c2bde3cd399fbc91380

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79548 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58557 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32925 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84127 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30639 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81682 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67650 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6836 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62209 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20068 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82619 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52268 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72472 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42517 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49614 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26634 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29056 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70736 "") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27093 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85538 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6814 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4752 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70457 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6979 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68334 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69702 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13731 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12623 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12304 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6766 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12454 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6663 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10142 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8461 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

axiixc commented on Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift:
```diff
@@ -0,0 +1,61 @@
+
+extension View {
+    @_spi(Private)
+    public func webViewAllowsBackForwardNavigationGestures(_ value: Bool = true) -> some View {
```
Did we ever decide whether we liked this spelling or `webViewBackForwardNavigationGestures(.enabled)` as the spelling? Or if we should leave off the default value? I do agree it’s kinda weird to have the no-arg spelling just restate the default value.

axiixc replied:
I guess one downside to a bespoke type would be that you can’t directly bind it to something like `Toggle(isOn:)`, though I’m not sure how useful that is in practice anyways.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -120,20 +125,125 @@ public class WebPage_v0 {
         ]
     }
 
+    public static func handlesURLScheme(_ scheme: String) -> Bool {
```
`static` should probably go at the top of the type, like we would for a class method in ObjC.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -120,20 +125,125 @@ public class WebPage_v0 {
     }
 
+    @discardableResult
+    public func load(_ data: Data, mimeType: String, characterEncoding: String.Encoding, baseURL: URL) -> NavigationID? {
```
**For later:** I wonder if it’s better to use `mimeType: String` or `contentType: UTType` here. We didn’t have a strongly typed framework for this back when the `WKWebView` API was added, but I’m also unsure if that’d be a usability regression for providing parameters like `text/plain;charset=UTF-8`.

aprotyas replied:
I do agree that we should eventually consider a `UTType` argument here.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -120,20 +125,125 @@ public class WebPage_v0 {
+    public func load(_ data: Data, mimeType: String, characterEncoding: String.Encoding, baseURL: URL) -> NavigationID? {
+        let cfEncoding = CFStringConvertNSStringEncodingToEncoding(characterEncoding.rawValue)
+        guard cfEncoding != kCFStringEncodingInvalidId else {
+            preconditionFailure("\(characterEncoding) is not a valid character encoding")
```
Are the supported encodings fixed by the OS? Making this a precondition means that we’re calling it a programmer error, but I’m wondering how easy it is for developers to get right.

Put another way, would we throw an `NSException` for this in ObjC?

rr-codes replied:
In actuality, I don't think this case would ever happen since devs would 99% use one of the static properties of String.Encoding, so I just put that precondition there just in case.

cc @beidson for the actual answers to the questions though

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -120,20 +125,125 @@ public class WebPage_v0 {
 
+    @discardableResult
+    public func load(_ request: URLRequest, allowingReadAccessTo readAccessURL: URL) -> NavigationID? {
+        // `WKWebView` annotates this method as returning non-nil, but it may return nil.
```
Also for later, but this seems worth fixing more generally for Swift, even if it has source breaking implications.

aprotyas replied:
Would this also lead to issues like this? https://github.com/WebKit/WebKit/pull/37519

rr-codes replied:
nope, since it's not async

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -120,20 +125,125 @@ public class WebPage_v0 {
+    }
+
+    public func callAsyncJavaScript(_ functionBody: String, arguments: [String : Any] = [:], in frame: FrameInfo? = nil, contentWorld: WKContentWorld? = nil) async throws -> Any? {
+        try await backingWebView.callAsyncJavaScript(functionBody, arguments: arguments, in: frame?.wrapped, contentWorld: contentWorld ?? .defaultClient)
```
Can we just use `.defaultClient` as the default argument value instead of making it optional? (I can’t remember if there was a reason we didn’t do that previously.)

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -120,20 +125,125 @@ public class WebPage_v0 {
+                    return
+                }
+
+                guard let image else {
```
This isn’t really any different than force unwrapping, just more verbose. Given that all the invariants that would trigger that are in our own code, it’s probably fine to just use `!`. Though, did we not produce an `async` overload of this on `WKWebView` that we could just use instead?

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -120,20 +125,125 @@ public class WebPage_v0 {
+#endif
+
+    public func pdf(configuration: WKPDFConfiguration? = nil) async throws -> Data {
+        try await backingWebView.pdf(configuration: configuration ?? .init())
```
I was almost going to suggest making the configuration non-optional above, but since `nil` would flow through all the layers it seemed unnecessary. This give some some pause since there really isn’t any reason to make that `= .init()` just the default value (it won’t incur any runtime penalty, default values are inlined into the client binary).

That does make me kinda wish we’d have consistent optionality, but I think even without this should just use a default value.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -120,20 +125,125 @@ public class WebPage_v0 {
+    }
+
+    public func webArchiveData() async throws -> Data {
+        try await withCheckedThrowingContinuation { continuation in
```
This doesn’t appear to actually throw. Do we just want to reserve that future flexibility?

rr-codes replied:
doesn't it throw? since I'm returning a Result<T, E> to the continuation?

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebView.swift:
```diff
@@ -103,7 +104,16 @@ fileprivate struct WebViewRepresentable {
+
+        platformView.webView = webView
+
+        webView.allowsBackForwardNavigationGestures = environment.webViewAllowsBackForwardNavigationGestures
```
Are we checking somewhere whether these values are actually changes. As in, this won’t kick off unnecessary work? This method could be called semi-frequently (mentally I’d categorize it like `-layoutSubviews` even though that’s not exactly true), so we should just be mindful of performance is all.

rr-codes replied:
the underlying setters in WKWebView have early returns in the case where the values don't change

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+FrameInfo.swift:
I don’t know if there’s a true convention on this, but we’ve taken to naming nested types as `Foo.Bar.swift` and using `+` to mean “breaking this up” or “extensions of”. WebKit never really go into the `+` naming for ObjC categories, so I’m not sure what kind of precedent to set here.

axiixc commented on Tools/MiniBrowser/SwiftUI/ContentView.swift:
```diff
@@ -26,15 +26,56 @@
 
     var body: some View {
         WebView_v0(viewModel.page)
+            .webViewAllowsBackForwardNavigationGestures()
```
Kinda related to the above, but since this doesn’t do anything to change the default I’m not sure whether it’s useful to include. We could have a `Toggle` somewhere to properly exercise it maybe?

rr-codes replied:
it does do something; if it's excluded, the default value is false. but the default value of the parameter is true, so this sets the preference to true from false

rr-codes committed:
[SwiftUI] Add more of the main functionality to WebPage and WebView
https://bugs.webkit.org/show_bug.cgi?id=283653
rdar://140507710

Reviewed by Abrar Rahman Protyasha.

This adds most of the missing capabilities to WebPage and WebView.

* Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift: Added.
(View.webViewAllowsBackForwardNavigationGestures(_:)):
(View.webViewAllowsLinkPreview(_:)):
(View.webViewAllowsTabFocusingLinks(_:)):
(View.webViewAllowsTextInteraction(_:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+FrameInfo.swift: Added.
(FrameInfo.isMainFrame):
(FrameInfo.request):
(FrameInfo.securityOrigin):
(FrameInfo.wrapped):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.isWritingToolsActive):
(WebPage_v0.handlesURLScheme(_:)):
(WebPage_v0.load(_:mimeType:characterEncoding:baseURL:)):
(WebPage_v0.load(_:allowingReadAccessTo:)):
(WebPage_v0.loadSimulatedRequest(_:response:responseData:)):
(WebPage_v0.loadSimulatedRequest(_:responseHTML:)):
(WebPage_v0.reload(_:)):
(WebPage_v0.stopLoading):
(WebPage_v0.callAsyncJavaScript(_:arguments:in:contentWorld:)):
(WebPage_v0.snapshot(_:)):
(WebPage_v0.pdf(_:)):
(WebPage_v0.webArchiveData):
* Source/WebKit/UIProcess/API/Swift/WebView.swift:
(WebViewRepresentable.updatePlatformView(_:context:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/MiniBrowser/SwiftUI/ContentView.swift:
(UnifiedBar.body):
(ContentView.body):
(ContentView.urlFieldIsFocused): Deleted.
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift:
(WebPageTests.javaScriptEvaluation):

Canonical link: https://commits.webkit.org/287428@main
